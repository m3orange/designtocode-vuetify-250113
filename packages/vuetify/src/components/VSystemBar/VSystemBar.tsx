// Styles
import './VSystemBar.sass'

// Composables
import { makeElevationProps, useElevation } from '@/composables/elevation'
import { makeRoundedProps, useRounded } from '@/composables/rounded'
import { makeTagProps } from '@/composables/tag'
import { makeThemeProps, provideTheme } from '@/composables/theme'
import { useBackgroundColor } from '@/composables/color'
import { makeLayoutItemProps, useLayoutItem } from '@/composables/layout'

// Utilities
import { defineComponent } from '@/util'
import { computed, ref, toRef } from 'vue'
import { provideDefaults } from '@/composables/defaults'

export const VSystemBar = defineComponent({
  name: 'VSystemBar',

  props: {
    color: String,
    window: Boolean,
    height: [Number, String],

    ...makeElevationProps(),
    ...makeLayoutItemProps(),
    ...makeRoundedProps(),
    ...makeTagProps(),
    ...makeThemeProps(),
  },

  setup (props, { slots }) {
    const { themeClasses } = provideTheme(props)
    const { backgroundColorClasses, backgroundColorStyles } = useBackgroundColor(toRef(props, 'color'))
    const { elevationClasses } = useElevation(props)
    const { roundedClasses } = useRounded(props)
    const height = computed(() => props.height ?? props.window ? 32 : 24)
    const { layoutItemStyles } = useLayoutItem({
      id: props.name,
      priority: computed(() => parseInt(props.priority, 10)),
      position: ref('top'),
      layoutSize: height,
      elementSize: height,
      active: computed(() => true),
      absolute: toRef(props, 'absolute'),
    })

    provideDefaults({
      VBtn: {
        variant: 'text',
        density: 'compact',
      },
    }, { scoped: true })

    return () => (
      <props.tag
        class={[
          'v-system-bar',
          { 'v-system-bar--window': props.window },
          themeClasses.value,
          backgroundColorClasses.value,
          elevationClasses.value,
          roundedClasses.value,
        ]}
        style={[
          backgroundColorStyles.value,
          layoutItemStyles.value,
        ]}
        v-slots={ slots }
      />
    )
  },
})
