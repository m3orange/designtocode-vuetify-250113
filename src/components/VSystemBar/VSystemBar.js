require('../../stylus/components/_system-bars.styl')

import Applicationable from '../../mixins/applicationable'
import Colorable from '../../mixins/colorable'
import Themeable from '../../mixins/themeable'

export default {
  name: 'v-system-bar',

  mixins: [Applicationable, Colorable, Themeable],

  props: {
    lightsOut: Boolean,
    status: Boolean,
    window: Boolean
  },

  computed: {
    classes () {
      return this.addBackgroundColorClassChecks(Object.assign({
        'system-bar--lights-out': this.lightsOut,
        'system-bar--fixed': this.app,
        'system-bar--status': this.status,
        'system-bar--window': this.window
      }, this.themeClasses))
    },
    computedHeight () {
      return this.window ? 32 : 24
    }
  },

  watch: {
    window () {
      this.updateApplication()
    }
  },

  methods: {
    updateApplication () {
      if (this.app) this.$vuetify.application.bar = this.computedHeight
    }
  },

  destroyed () {
    if (this.app) this.$vuetify.application.bar = 0
  },

  render (h) {
    const data = {
      staticClass: 'system-bar',
      'class': this.classes,
      style: {
        height: `${this.computedHeight}px`
      }
    }

    return h('div', data, this.$slots.default)
  }
}
