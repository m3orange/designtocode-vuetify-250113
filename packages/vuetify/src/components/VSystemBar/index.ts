// Styles
import './VSystemBar.sass'

import VSystemBar from './VSystemBar'

export { VSystemBar }
export default VSystemBar
