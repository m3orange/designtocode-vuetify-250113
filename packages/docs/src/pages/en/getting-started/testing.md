---
nav: Testing
meta:
  title: Testing
  description: Details for v3 beta release - faq, changes, and upgrading.
  keywords: test, testing, vitest, jest, vue 3, vuetify 3
---

<script setup>
  import { ref } from 'vue'

  const tab = ref('testing-library')
</script>

# Testing

Introduction here

## End-to-end testing

### Cypress

TBW

## Unit testing

We recommend [vitest](https://vitest.dev/) for your unit testing needs, if you are running a vite based project.

### Vitest

Start by installing `vitest` and the `jsdom` environment.

```bash
yarn -D vitest jsdom
```

Then configure vitest using the `test` object in the vite config. Make sure to add `vuetify` as an inline dependency.

```ts { data-resource="vite.config.ts" }
...

export default defineConfig({
  ...,
  test: {
    globals: true,
    environment: 'jsdom',
    deps: {
      inline: ['vuetify']
    }
  }
})
```

That should be enough to run tests. Go to the [Writing tests](#writing-tests) section to see how to render and test Vuetify components

### Jest

If you already have an existing jest setup that you want to continue using, have a look at [vite-jest](https://github.com/sodatea/vite-jest), which offers first-class Vite integration for jest.

### Writing tests

Here is a simple example component with a **VBtn** component.

```html { data-resource="Hello.vue" }
<script setup lang="ts">
    import { computed, ref } from 'vue'
    import { VBtn } from 'vuetify/components'

    const props = defineProps<{ count: number }>()
    const times = ref(2)
    const result = computed(() => props.count * times.value)

    defineExpose(props)
</script>

<template>
    <div>{{ count }} x {{ times }} = {{ result }}</div>
    <VBtn @click="times += 1">
    x1
    </VBtn>
</template>
```

To render Vue components in unit tests, you can use either [@testing-library/vue](https://github.com/testing-library/vue-testing-library) or [@vue/test-utils](https://github.com/vuejs/test-utils). You can find a good comparison of the two in the [Vue documentation](https://vuejs.org/guide/scaling-up/testing.html#mounting-libraries).

There is not much difference between the two when it comes to mounting Vuetify components. In the examples below we have created a helper function that registers Vuetify as a global plugin.

<v-tabs v-model="tab" color="primary">
  <v-tab value="testing-library" variant="plain">@testing-library/vue</v-tab>
  <v-tab value="test-utils" variant="plain">@vue/test-utils</v-tab>
</v-tabs>

<v-window v-model="tab">
  <v-window-item value="testing-library">

```ts { data-resource="Hello.spec.ts" }
import { render, fireEvent } from '@testing-library/vue'
import { test, expect } from 'vitest'
import { createVuetify } from 'vuetify'
import Hello from './Hello.vue'

function renderWithVuetify (component, options) {
    const vuetify = createVuetify()
    return render(component, {
        ...options,
        global: {
            plugins: [vuetify]
        },
    })
}

test('mount component', async () => {
  expect(Hello).toBeTruthy()

  const wrapper = renderWithVuetify(Hello, {
    props: {
      count: 4,
    },
  })

  expect(wrapper.queryByText('4 x 2 = 8')).toBeTruthy()
  expect(wrapper.html()).toMatchSnapshot()

  const button = wrapper.queryByText('x1')!

  await fireEvent.click(button)

  expect(wrapper.queryByText('4 x 3 = 12')).toBeTruthy()

  await fireEvent.click(button)

  expect(wrapper.queryByText('4 x 4 = 16')).toBeTruthy()
})
```

  </v-window-item>
  <v-window-item value="test-utils">

```ts { data-resource="Hello.spec.ts" }
import { mount } from '@vue/test-utils'
import { test, expect } from 'vitest'
import { createVuetify } from 'vuetify'
import Hello from './Hello.vue'

function mountWithVuetify (component, options) {
    const vuetify = createVuetify()
    return mount(component, {
        ...options,
        global: {
            plugins: [vuetify]
        },
    })
}

test('mount component', async () => {
  expect(Hello).toBeTruthy()

  const wrapper = mountWithVuetify(Hello, {
    props: {
      count: 4,
    },
  })

  expect(wrapper.text()).toContain('4 x 2 = 8')
  expect(wrapper.html()).toMatchSnapshot()

  await wrapper.get('.v-btn').trigger('click')

  expect(wrapper.text()).toContain('4 x 3 = 12')

  await wrapper.get('.v-btn').trigger('click')

  expect(wrapper.text()).toContain('4 x 4 = 16')
})
```

  </v-window-item>
</v-window>





