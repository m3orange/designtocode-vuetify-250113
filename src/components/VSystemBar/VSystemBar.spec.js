import { test } from '~util/testing'
import { mount } from 'avoriaz'
import VSystemBar from '~components/VSystemBar'

test('VSystemBar.vue', () => {
  it('should render a colored toolbar', () => {
    const wrapper = mount(VSystemBar, {
      propsData: {
        color: 'blue lighten-1'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
