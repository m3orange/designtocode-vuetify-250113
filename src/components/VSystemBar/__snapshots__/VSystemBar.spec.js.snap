// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VSystemBar.vue should render a colored toolbar 1`] = `

<div class="system-bar blue lighten-1"
     style="height: 24px;"
>
</div>

`;
